dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,655 @@
+   *
+   * @return {Promise<void>} A promise that resolves when there are no more
+   * pending writes.
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   *
+   * bulkWriter.create(documentRef, {foo: 'bar'});
+   * bulkWriter.update(documentRef2, {foo: 'bar'});
+   * bulkWriter.delete(documentRef3);
+   * await flush().then(() => {
+   *   console.log('Successfully executed all writes');
+   * });
+   */
+  flush(): Promise<void> {
+    this.verifyNotClosed();
+    return this._flush();
+  }
+
+  /**
+   * Commits all enqueued writes and marks the BulkWriter instance as closed.
+   *
+   * After calling `close()`, calling any method wil throw an error.
+   *
+   * Returns a Promise that resolves when there are no more pending writes. It
+   * never fails. Calling this method will send all requests. The promise
+   * resolves immediately if there are no pending writes.
+   *
+   * @return {Promise<void>} A promise that resolves when there are no more
+   * pending writes.
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   *
+   * bulkWriter.create(documentRef, {foo: 'bar'});
+   * bulkWriter.update(documentRef2, {foo: 'bar'});
+   * bulkWriter.delete(documentRef3);
+   * await close().then(() => {
+   *   console.log('Successfully executed all writes');
+   * });
+   */
+  close(): Promise<void> {
+    this.verifyNotClosed();
+    this.closed = true;
+    return this._flush();
+  }
+
+  private async _flush(): Promise<void> {
+    const trackedBatches = this.batchQueue;
+    const writePromises = trackedBatches.map(batch => batch.awaitBatch());
+    this.sendBatches();
+    await Promise.all(writePromises);
+  }
+
+  private verifyNotClosed(): void {
+    if (this.closed) {
+      throw new Error('BulkWriter has already been closed.');
+    }
+  }
+
+  /**
+   * Gradually ramps up the maximum number of operations per second by 50%
+   * every 5 minutes in line with the 500/50/5 rule.
+   */
+  private rampMaxOps(): void {
+    delayExecution(() => {
+      this.maxOpsPerSecond = this.maxOpsPerSecond * 1.5;
+      this.rampMaxOps();
+    }, 5 * 60 * 1000);
+  }
+
+  /**
+   * Return the first eligible batch that can hold a write to the provided
+   * reference, or creates one if no eligible batches are found.
+   */
+  private getEligibleBatch(ref: DocumentReference): BulkCommitBatch {
+    let batch = null;
+    let logWarning = this.batchQueue.length > 0;
+    for (const bulkBatch of this.batchQueue) {
+      if (!bulkBatch.refsInBatch.has(ref)) {
+        logWarning = false;
+      }
+      if (bulkBatch.canAddDoc(ref)) {
+        batch = bulkBatch;
+      }
+    }
+
+    if (batch === null) {
+      if (logWarning) {
+        console.warn(
+          '[BulkWriter]',
+          'Writing to the same document multiple times will slow down BulkWriter. ' +
+            'Write to unique documents in order to maximize throughput.'
+        );
+      }
+      return this.createNewBatch();
```

Comment:
If we create a new batch, we should immediately send the the old batch. "flush()" is only an optimization in case the user wants to send off a smaller batch.